src/types/weakref/reference.rs

```diff
@@ -0,0 +1,1200 @@
+use crate::err::PyResult;
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::type_object::{PyTypeCheck, PyTypeInfo};
+use crate::types::any::PyAnyMethods;
+use crate::{ffi, Borrowed, Bound, PyAny, PyNativeType, Python, ToPyObject};
+
+/// Represents a Python `weakref.ReferenceType`.
+///
+/// In Python this is created by calling `weakref.ref`.
+#[repr(transparent)]
+pub struct PyWeakRef(PyAny);
+
+pyobject_native_type!(
+    PyWeakRef,
+    ffi::PyWeakReference,
+    pyobject_native_static_type_object!(ffi::_PyWeakref_RefType),
+    #module=Some("weakref"),
+    #checkfunction=ffi::PyWeakref_CheckRefExact
+);
+
+impl PyWeakRef {
+    /// Deprecated form of [`PyWeakRef::new_bound`].
+    #[inline]
+    #[track_caller]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyWeakRef::new` will be replaced by `PyWeakRef::new_bound` in a future PyO3 version"
+        )
+    )]
+    pub fn new<T>(py: Python<'_>, object: T) -> PyResult<&'_ PyWeakRef>
+    where
+        T: ToPyObject,
+    {
+        Self::new_bound(py, object).map(Bound::into_gil_ref)
+    }
+
+    /// Constructs a new Weak Reference (`weakref.ref`/`weakref.ReferenceType`) for the given object.
+    ///
+    /// Returns a `TypeError` if `object` is not weak referenceable (Most native types and PyClasses without `weakref` flag).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyWeakRef;
+    ///
+    /// #[pyclass(weakref)]
+    /// struct Foo { /* fields omitted */ }
+    ///
+    /// # fn main() -> PyResult<()> {
+    /// Python::with_gil(|py| {
+    ///     let foo = Bound::new(py, Foo {})?;
+    ///     let weakref = PyWeakRef::new_bound(py, foo.clone())?;
+    ///     assert!(
+    ///         // In normal situations where a direct `Bound<'py, Foo>` is required use `upgrade::<Foo>`
+    ///         weakref.get_object()
+    ///             .map_or(false, |obj| obj.is(&foo))
+    ///     );
+    ///
+    ///     let weakref2 = PyWeakRef::new_bound(py, foo.clone())?;
+    ///     assert!(weakref.is(&weakref2));
+    ///
+    ///     drop(foo);
+    ///
+    ///     assert!(weakref.get_object().is_none());
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    #[track_caller]
+    pub fn new_bound<T>(py: Python<'_>, object: T) -> PyResult<Bound<'_, PyWeakRef>>
+    where
+        T: ToPyObject,
+    {
+        unsafe {
+            Bound::from_owned_ptr_or_err(
+                py,
+                ffi::PyWeakref_NewRef(object.to_object(py).as_ptr(), ffi::Py_None()),
+            )
+            .map(|obj| obj.downcast_into_unchecked())
```

Comment:
Can use our internal `PyResultExt` trait to skip the `.map` here and in other locations:

```suggestion
            .downcast_into_unchecked()
```

Reply:
I think I changed all relevant occurrences.